pkg/kubelet/cadvisor/cadvisor_windows.go

```diff
@@ -22,16 +22,19 @@ import (
 	"github.com/google/cadvisor/events"
 	cadvisorapi "github.com/google/cadvisor/info/v1"
 	cadvisorapiv2 "github.com/google/cadvisor/info/v2"
+	"k8s.io/kubernetes/pkg/kubelet/winstats"
 )
 
 type cadvisorClient struct {
+	winStatsClient *winstats.Client
 }
 
 var _ Interface = new(cadvisorClient)
 
 // New creates a cAdvisor and exports its API on the specified port if port > 0.
 func New(address string, port uint, runtime string, rootPath string) (Interface, error) {
-	return &cadvisorClient{}, nil
+	client, err := winstats.NewClient(winstats.NewPerfCounterNodeStatsClient())
```

Comment:
The reason I did this is for the [unit tests](https://github.com/bobbypage/kubernetes/blob/2c8d2794bb1652a9d282df9d7e3abc2c7d7bcb8b/pkg/kubelet/winstats/winstats_test.go). I wasn't able to create windows unit tests (tests don't build on Windows), so I created tests that work on linux. Since some of the code relies on windows specific perf counters, I made it so you can pass a node stats client (in the tests I created a mock that is passed into `winstats.NewClient`). That is the reason I exported `PerfCounterNodeStatsClient`. Let me know if this approach makes sense...

Reply:
Since the test can access the unexported types here, you can create the client using `c := &cadvisorClient{&fakeWinNodeStatsClient{}}`.

Why can't the tests be built on Windows?

Reply:
+1
You don't need to export the type just for the unit tests. We have plenty of unit tests that do what @yguo0905 suggested.

The fact that tests don't build on windows seems concerning. Can you paste the error message?

Reply:
I looked into building the tests on Windows. There a lot of errors with missing packages. For example, `apimachinery` is under staging directory and I think since  `make` does not seem to work on windows, the packages are not put in the correct location. Here is an example: 
```
PS C:\code\Go\src\k8s.io\kubernetes> go test -v k8s.io/kubernetes/pkg/kubelet/
pkg\controller\controller_ref_manager.go:24:2: cannot find package "k8s.io/api/apps/v1beta1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\api\apps\v1beta1 (vendor tree)
        C:\tools\go\src\k8s.io\api\apps\v1beta1 (from $GOROOT)
        C:\code\Go\src\k8s.io\api\apps\v1beta1 (from $GOPATH)
pkg\controller\client_builder.go:23:2: cannot find package "k8s.io/api/authentication/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\api\authentication\v1 (vendor tree)
        C:\tools\go\src\k8s.io\api\authentication\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\api\authentication\v1 (from $GOPATH)
pkg\kubelet\certificate\certificate_manager.go:32:2: cannot find package "k8s.io/api/certificates/v1beta1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\api\certificates\v1beta1 (vendor tree)
        C:\tools\go\src\k8s.io\api\certificates\v1beta1 (from $GOROOT)
        C:\code\Go\src\k8s.io\api\certificates\v1beta1 (from $GOPATH)
pkg\kubelet\active_deadline.go:23:2: cannot find package "k8s.io/api/core/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\api\core\v1 (vendor tree)
        C:\tools\go\src\k8s.io\api\core\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\api\core\v1 (from $GOPATH)
pkg\controller\controller_ref_manager.go:26:2: cannot find package "k8s.io/api/extensions/v1beta1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\api\extensions\v1beta1 (vendor tree)
        C:\tools\go\src\k8s.io\api\extensions\v1beta1 (from $GOROOT)
        C:\code\Go\src\k8s.io\api\extensions\v1beta1 (from $GOPATH)
pkg\volume\util\util.go:27:2: cannot find package "k8s.io/api/storage/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\api\storage\v1 (vendor tree)
        C:\tools\go\src\k8s.io\api\storage\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\api\storage\v1 (from $GOPATH)
pkg\api\validation\validation.go:34:2: cannot find package "k8s.io/apimachinery/pkg/api/equality" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\api\equality (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\api\equality (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\api\equality (from $GOPATH)
pkg\kubelet\kubelet_node_status.go:30:2: cannot find package "k8s.io/apimachinery/pkg/api/errors" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\api\errors (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\api\errors (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\api\errors (from $GOPATH)
pkg\controller\controller_utils.go:30:2: cannot find package "k8s.io/apimachinery/pkg/api/meta" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\api\meta (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\api\meta (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\api\meta (from $GOPATH)
pkg\kubelet\kubelet_node_status.go:31:2: cannot find package "k8s.io/apimachinery/pkg/api/resource" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\api\resource (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\api\resource (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\api\resource (from $GOPATH)
pkg\api\validation\validation.go:36:2: cannot find package "k8s.io/apimachinery/pkg/api/validation" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\api\validation (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\api\validation (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\api\validation (from $GOPATH)
pkg\api\register.go:22:2: cannot find package "k8s.io/apimachinery/pkg/apimachinery/announced" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\apimachinery\announced (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\apimachinery\announced (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\apimachinery\announced (from $GOPATH)
pkg\api\register.go:23:2: cannot find package "k8s.io/apimachinery/pkg/apimachinery/registered" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\apimachinery\registered (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\apimachinery\registered (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\apimachinery\registered (from $GOPATH)
pkg\api\types.go:21:2: cannot find package "k8s.io/apimachinery/pkg/apis/meta/internalversion" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\apis\meta\internalversion (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\apis\meta\internalversion (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\apis\meta\internalversion (from $GOPATH)
pkg\kubelet\kubelet.go:40:2: cannot find package "k8s.io/apimachinery/pkg/apis/meta/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\apis\meta\v1 (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\apis\meta\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\apis\meta\v1 (from $GOPATH)
pkg\api\validation\validation.go:39:2: cannot find package "k8s.io/apimachinery/pkg/apis/meta/v1/validation" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\apis\meta\v1\validation (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\apis\meta\v1\validation (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\apis\meta\v1\validation (from $GOPATH)
pkg\kubelet\kubelet_node_status.go:33:2: cannot find package "k8s.io/apimachinery/pkg/conversion" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\conversion (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\conversion (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\conversion (from $GOPATH)
pkg\kubelet\kubelet.go:41:2: cannot find package "k8s.io/apimachinery/pkg/fields" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\fields (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\fields (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\fields (from $GOPATH)
pkg\kubelet\kubelet.go:42:2: cannot find package "k8s.io/apimachinery/pkg/labels" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\labels (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\labels (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\labels (from $GOPATH)
pkg\apis\componentconfig\helpers.go:26:2: cannot find package "k8s.io/apimachinery/pkg/runtime" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\runtime (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\runtime (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\runtime (from $GOPATH)
pkg\apis\componentconfig\register.go:21:2: cannot find package "k8s.io/apimachinery/pkg/runtime/schema" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\runtime\schema (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\runtime\schema (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\runtime\schema (from $GOPATH)
pkg\api\register.go:27:2: cannot find package "k8s.io/apimachinery/pkg/runtime/serializer" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\runtime\serializer (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\runtime\serializer (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\runtime\serializer (from $GOPATH)
pkg\api\helper\helpers.go:29:2: cannot find package "k8s.io/apimachinery/pkg/selection" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\selection (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\selection (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\selection (from $GOPATH)
pkg\kubelet\kubelet.go:43:2: cannot find package "k8s.io/apimachinery/pkg/types" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\types (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\types (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\types (from $GOPATH)
pkg\kubelet\active_deadline.go:24:2: cannot find package "k8s.io/apimachinery/pkg/util/clock" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\clock (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\clock (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\clock (from $GOPATH)
pkg\api\validation\validation.go:41:2: cannot find package "k8s.io/apimachinery/pkg/util/diff" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\diff (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\diff (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\diff (from $GOPATH)
pkg\kubelet\kubelet_volumes.go:25:2: cannot find package "k8s.io/apimachinery/pkg/util/errors" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\errors (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\errors (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\errors (from $GOPATH)
pkg\kubelet\server\portforward\httpstream.go:28:2: cannot find package "k8s.io/apimachinery/pkg/util/httpstream" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\httpstream (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\httpstream (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\httpstream (from $GOPATH)
pkg\kubelet\server\portforward\httpstream.go:29:2: cannot find package "k8s.io/apimachinery/pkg/util/httpstream/spdy" in any of:        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\httpstream\spdy (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\httpstream\spdy (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\httpstream\spdy (from $GOPATH)
pkg\api\types.go:26:2: cannot find package "k8s.io/apimachinery/pkg/util/intstr" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\intstr (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\intstr (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\intstr (from $GOPATH)
pkg\kubelet\kubelet_node_status.go:35:2: cannot find package "k8s.io/apimachinery/pkg/util/net" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\net (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\net (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\net (from $GOPATH)
pkg\api\v1\generate.go:23:2: cannot find package "k8s.io/apimachinery/pkg/util/rand" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\rand (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\rand (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\rand (from $GOPATH)
pkg\kubelet\server\streaming\server.go:34:2: cannot find package "k8s.io/apimachinery/pkg/util/remotecommand" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\remotecommand (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\remotecommand (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\remotecommand (from $GOPATH)
pkg\kubelet\kubelet.go:45:2: cannot find package "k8s.io/apimachinery/pkg/util/runtime" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\runtime (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\runtime (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\runtime (from $GOPATH)
pkg\kubelet\kubelet.go:46:2: cannot find package "k8s.io/apimachinery/pkg/util/sets" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\sets (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\sets (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\sets (from $GOPATH)
pkg\controller\controller_utils.go:38:2: cannot find package "k8s.io/apimachinery/pkg/util/strategicpatch" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\strategicpatch (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\strategicpatch (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\strategicpatch (from $GOPATH)
pkg\kubelet\rkt\rkt.go:48:2: cannot find package "k8s.io/apimachinery/pkg/util/uuid" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\uuid (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\uuid (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\uuid (from $GOPATH)
pkg\kubelet\kubelet_pods.go:42:2: cannot find package "k8s.io/apimachinery/pkg/util/validation" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\validation (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\validation (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\validation (from $GOPATH)
pkg\kubelet\kubelet_pods.go:43:2: cannot find package "k8s.io/apimachinery/pkg/util/validation/field" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\validation\field (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\validation\field (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\validation\field (from $GOPATH)
pkg\kubelet\kubelet.go:47:2: cannot find package "k8s.io/apimachinery/pkg/util/wait" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\wait (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\wait (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\wait (from $GOPATH)
pkg\kubelet\config\common.go:29:2: cannot find package "k8s.io/apimachinery/pkg/util/yaml" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\util\yaml (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\util\yaml (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\util\yaml (from $GOPATH)
pkg\version\version.go:23:2: cannot find package "k8s.io/apimachinery/pkg/version" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\version (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\version (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\version (from $GOPATH)
pkg\controller\client_builder.go:29:2: cannot find package "k8s.io/apimachinery/pkg/watch" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apimachinery\pkg\watch (vendor tree)
        C:\tools\go\src\k8s.io\apimachinery\pkg\watch (from $GOROOT)
        C:\code\Go\src\k8s.io\apimachinery\pkg\watch (from $GOPATH)
pkg\quota\evaluator\core\persistent_volume_claims.go:29:2: cannot find package "k8s.io/apiserver/pkg/admission" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\admission (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\admission (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\admission (from $GOPATH)
pkg\serviceaccount\jwt.go:28:2: cannot find package "k8s.io/apiserver/pkg/authentication/authenticator" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\authentication\authenticator (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\authentication\authenticator (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\authentication\authenticator (from $GOPATH)
pkg\controller\client_builder.go:30:2: cannot find package "k8s.io/apiserver/pkg/authentication/serviceaccount" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\authentication\serviceaccount (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\authentication\serviceaccount (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\authentication\serviceaccount (from $GOPATH)
pkg\serviceaccount\jwt.go:30:2: cannot find package "k8s.io/apiserver/pkg/authentication/user" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\authentication\user (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\authentication\user (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\authentication\user (from $GOPATH)
pkg\kubelet\server\auth.go:27:2: cannot find package "k8s.io/apiserver/pkg/authorization/authorizer" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\authorization\authorizer (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\authorization\authorizer (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\authorization\authorizer (from $GOPATH)
pkg\features\kube_features.go:20:2: cannot find package "k8s.io/apiserver/pkg/features" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\features (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\features (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\features (from $GOPATH)
pkg\kubelet\server\server.go:50:2: cannot find package "k8s.io/apiserver/pkg/server/healthz" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\server\healthz (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\server\healthz (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\server\healthz (from $GOPATH)
pkg\kubelet\server\portforward\websocket.go:33:2: cannot find package "k8s.io/apiserver/pkg/server/httplog" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\server\httplog (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\server\httplog (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\server\httplog (from $GOPATH)
pkg\kubelet\configmap\configmap_manager.go:26:2: cannot find package "k8s.io/apiserver/pkg/storage/etcd" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\storage\etcd (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\storage\etcd (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\storage\etcd (from $GOPATH)
pkg\kubelet\kubelet.go:48:2: cannot find package "k8s.io/apiserver/pkg/util/feature" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\util\feature (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\util\feature (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\util\feature (from $GOPATH)
cmd\kubelet\app\options\options.go:26:2: cannot find package "k8s.io/apiserver/pkg/util/flag" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\util\flag (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\util\flag (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\util\flag (from $GOPATH)
pkg\kubelet\server\server.go:52:2: cannot find package "k8s.io/apiserver/pkg/util/flushwriter" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\util\flushwriter (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\util\flushwriter (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\util\flushwriter (from $GOPATH)
pkg\kubelet\server\portforward\portforward.go:26:2: cannot find package "k8s.io/apiserver/pkg/util/wsstream" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\apiserver\pkg\util\wsstream (vendor tree)
        C:\tools\go\src\k8s.io\apiserver\pkg\util\wsstream (from $GOROOT)
        C:\code\Go\src\k8s.io\apiserver\pkg\util\wsstream (from $GOPATH)
pkg\quota\evaluator\core\configmap.go:23:2: cannot find package "k8s.io/client-go/informers" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\informers (vendor tree)
        C:\tools\go\src\k8s.io\client-go\informers (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\informers (from $GOPATH)
pkg\kubelet\kubelet.go:35:2: cannot find package "k8s.io/client-go/kubernetes" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\kubernetes (vendor tree)
        C:\tools\go\src\k8s.io\client-go\kubernetes (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\kubernetes (from $GOPATH)
pkg\kubelet\kubelet_resources.go:25:2: cannot find package "k8s.io/client-go/kubernetes/scheme" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\kubernetes\scheme (vendor tree)
        C:\tools\go\src\k8s.io\client-go\kubernetes\scheme (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\kubernetes\scheme (from $GOPATH)
pkg\controller\client_builder.go:33:2: cannot find package "k8s.io/client-go/kubernetes/typed/authentication/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\kubernetes\typed\authentication\v1 (vendor tree)
        C:\tools\go\src\k8s.io\client-go\kubernetes\typed\authentication\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\kubernetes\typed\authentication\v1 (from $GOPATH)
pkg\kubelet\certificate\certificate_manager.go:37:2: cannot find package "k8s.io/client-go/kubernetes/typed/certificates/v1beta1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\kubernetes\typed\certificates\v1beta1 (vendor tree)
        C:\tools\go\src\k8s.io\client-go\kubernetes\typed\certificates\v1beta1 (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\kubernetes\typed\certificates\v1beta1 (from $GOPATH)
pkg\kubelet\kubelet.go:50:2: cannot find package "k8s.io/client-go/kubernetes/typed/core/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\kubernetes\typed\core\v1 (vendor tree)
        C:\tools\go\src\k8s.io\client-go\kubernetes\typed\core\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\kubernetes\typed\core\v1 (from $GOPATH)
pkg\kubelet\kubelet.go:51:2: cannot find package "k8s.io/client-go/listers/core/v1" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\listers\core\v1 (vendor tree)
        C:\tools\go\src\k8s.io\client-go\listers\core\v1 (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\listers\core\v1 (from $GOPATH)
pkg\controller\client_builder.go:35:2: cannot find package "k8s.io/client-go/rest" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\rest (vendor tree)
        C:\tools\go\src\k8s.io\client-go\rest (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\rest (from $GOPATH)
pkg\kubelet\kubelet.go:52:2: cannot find package "k8s.io/client-go/tools/cache" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\tools\cache (vendor tree)
        C:\tools\go\src\k8s.io\client-go\tools\cache (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\tools\cache (from $GOPATH)
pkg\kubelet\active_deadline.go:25:2: cannot find package "k8s.io/client-go/tools/record" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\tools\record (vendor tree)
        C:\tools\go\src\k8s.io\client-go\tools\record (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\tools\record (from $GOPATH)
pkg\controller\controller_utils.go:44:2: cannot find package "k8s.io/client-go/tools/reference" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\tools\reference (vendor tree)
        C:\tools\go\src\k8s.io\client-go\tools\reference (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\tools\reference (from $GOPATH)
pkg\kubelet\kubelet_pods.go:44:2: cannot find package "k8s.io/client-go/tools/remotecommand" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\tools\remotecommand (vendor tree)
        C:\tools\go\src\k8s.io\client-go\tools\remotecommand (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\tools\remotecommand (from $GOPATH)
pkg\kubelet\certificate\certificate_manager.go:38:2: cannot find package "k8s.io/client-go/util/cert" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\util\cert (vendor tree)
        C:\tools\go\src\k8s.io\client-go\util\cert (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\util\cert (from $GOPATH)
pkg\kubelet\kubelet.go:54:2: cannot find package "k8s.io/client-go/util/flowcontrol" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\util\flowcontrol (vendor tree)
        C:\tools\go\src\k8s.io\client-go\util\flowcontrol (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\util\flowcontrol (from $GOPATH)
pkg\kubelet\kubelet.go:55:2: cannot find package "k8s.io/client-go/util/integer" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\util\integer (vendor tree)
        C:\tools\go\src\k8s.io\client-go\util\integer (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\util\integer (from $GOPATH)
pkg\controller\controller_utils.go:46:2: cannot find package "k8s.io/client-go/util/retry" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\util\retry (vendor tree)
        C:\tools\go\src\k8s.io\client-go\util\retry (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\util\retry (from $GOPATH)
plugin\pkg\scheduler\algorithm\predicates\predicates.go:34:2: cannot find package "k8s.io/client-go/util/workqueue" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\client-go\util\workqueue (vendor tree)
        C:\tools\go\src\k8s.io\client-go\util\workqueue (from $GOROOT)
        C:\code\Go\src\k8s.io\client-go\util\workqueue (from $GOPATH)
plugin\pkg\scheduler\algorithm\predicates\predicates.go:44:2: cannot find package "k8s.io/metrics/pkg/client/clientset_generated/clientset" in any of:
        C:\code\Go\src\k8s.io\kubernetes\vendor\k8s.io\metrics\pkg\client\clientset_generated\clientset (vendor tree)
        C:\tools\go\src\k8s.io\metrics\pkg\client\clientset_generated\clientset (from $GOROOT)
        C:\code\Go\src\k8s.io\metrics\pkg\client\clientset_generated\clientset (from $GOPATH)
PS C:\code\Go\src\k8s.io\kubernetes>
```

Reply:
I'm not familiar with windows build, but did you also get this error when building kubelet, or it's test only?

Reply:
In the `vendor` directory, there are [symbolic links](https://github.com/kubernetes/kubernetes/tree/v1.9.0-alpha.0/vendor/k8s.io) to the packages in the `staging` directory. I am not sure how/whether this works on windows,. You may want to check if they are there.  Here's the [script]( https://github.com/kubernetes/kubernetes/blob/v1.9.0-alpha.0/hack/godep-save.sh#L50) that (I think) creates these links.

I'd suggest asking folks from sig-windows for more advice.

Reply:
Thanks. I don't think building on Windows is really supported (all the make files and infrastructure rely on Linux tools). I think most of folks working on Windows specific stuff have been cross compiling from Linux to get Windows binaries.

Reply:
Also regarding exporting the type, I did not export the type for testing. The type for testing is `fakeWinNodeStatsClient` and is not exported. However, `PerfCounterNodeStatsClient` is exported because it is passed into `NewClient()` inside a different package `pkg/kubelet/cadvisor`/

Reply:
I think the idea is that we don't have to expose `PerfCounterNodeStatsClient`. I feel a little bit weird in this case to expose an unexported interface `winNodeStatsClient` in the public function `NewClient()`. I think `pkg/kubelet/cadvisor` can just call `winstats.NewClient()`.

Reply:
+1.

`perfCounterNodeStatsClient` should be initialized automatically in `winstats.NewClient()` or if you prefer, a new function `winstats.NewPerfCounterClient()` can be added. 
There is no need to export PerfCounter related types.

Reply:
Got it, I have updated the PR. I added a new `winstats.NewPerfCounterClient()` as you suggested. That way I can not export the type, and use a separate `winstats.NewClient()`, which can pass in a mock for testing.